src/buffer/out/UTextAdapter.cpp

```diff
@@ -0,0 +1,328 @@
+// at the start or end of the string, as appropriate.
+//
+// @param ut          the UText being accessed.
+// @param nativeIndex Requested index of the text to be accessed.
+// @param forward     If true, then the returned chunk must contain text starting from the index, so that start<=index<limit.
+//                    If false, then the returned chunk must contain text before the index, so that start<index<=limit.
+// @return            True if the requested index could be accessed. The chunk will contain the requested text.
+//                    False value if a chunk cannot be accessed (the requested index is out of bounds).
+static UBool U_CALLCONV utextAccess(UText* ut, int64_t nativeIndex, UBool forward) noexcept
+try
+{
+    if (nativeIndex < 0)
+    {
+        nativeIndex = 0;
+    }
+
+    auto neededIndex = nativeIndex;
+    if (!forward)
+    {
+        neededIndex--;
+    }
+
+    const auto& textBuffer = *static_cast<const TextBuffer*>(ut->context);
+    const auto range = accessRowRange(ut);
+    auto start = ut->chunkNativeStart;
+    auto limit = ut->chunkNativeLimit;
+    auto y = accessCurrentRow(ut);
+    std::wstring_view text;
+
+    if (neededIndex < start || neededIndex >= limit)
+    {
+        if (neededIndex < start)
+        {
+            do
+            {
+                --y;
+                if (y < range.begin)
+                {
+                    return false;
+                }
+
+                text = textBuffer.GetRowByOffset(y).GetText();
+                limit = start;
+                start -= text.size();
+            } while (neededIndex < start);
+        }
+        else
+        {
+            do
+            {
+                ++y;
+                if (y >= range.end)
+                {
+                    return false;
+                }
+
+                text = textBuffer.GetRowByOffset(y).GetText();
+                start = limit;
+                limit += text.size();
+            } while (neededIndex >= limit);
+        }
+
+        accessCurrentRow(ut) = y;
+        ut->chunkNativeStart = start;
+        ut->chunkNativeLimit = limit;
+        ut->chunkLength = gsl::narrow_cast<int32_t>(text.size());
+#pragma warning(suppress : 26490) // Don't use reinterpret_cast (type.1).
+        ut->chunkContents = reinterpret_cast<const char16_t*>(text.data());
+        ut->nativeIndexingLimit = ut->chunkLength;
+    }
+
+    auto offset = gsl::narrow_cast<int32_t>(nativeIndex - start);
+
+    // Don't leave the offset on a trailing surrogate pair. See U16_SET_CP_START.
+    // This assumes that the TextBuffer contains valid UTF-16 which may theoretically not be the case.
+    if (offset > 0 && offset < ut->chunkLength && U16_IS_TRAIL(til::at(ut->chunkContents, offset)))
+    {
+        offset--;
+    }
+
+    ut->chunkOffset = offset;
+    return true;
+}
+catch (...)
+{
+    return false;
+}
+
+// An excerpt from the ICU documentation:
+//
+// Extract text from a UText into a UChar buffer.

```

Comment:
Did you complete this one?

Reply:
No, I'll keep the code but remove it from the `utextFuncs` struct and add a comment that the correctness of the function hasn't been verified yet. I don't think it's worth it to finish that function, because we don't need it right now. But I do think it's worth keeping because the code has already been written and is probably 99% correct (if not 100%). I initially thought I had to implement it so that `uregex` can work, but it seems I was wrong about that. It uses the `ut->chunkContents` directly (via `UTEXT_NEXT32` and friends).